alpa/serialization.py

```diff
@@ -28,191 +24,154 @@
-        and ReplicatedDistributedArray in alpa. Also it will save np.ndarray and jax.xla.DeviceArray
-        into tensorstore for later distributed loading.
-        # TODO (zhongyinmin): copy all the safe-saving stuff from 
-        https://flax.readthedocs.io/en/latest/_modules/flax/training/checkpoints.html#save_checkpoint
+        If you just want to save a unparallelized model or the model is parallellized on a single node,
+        `ckpt_dir` should be a normal path on local disk, and the `local_cache_dir` should be None.
 
         Args:
-           ckpt_dir: str or pathlib-like path to store checkpoint directories in.
-           target: serializable flax object, usually a trainState
+           ckpt_dir: the directory where this checkpoint will be saved.
+           target: serializable flax object, usually a trainState.
            step: training step number or other metric number
+           local_cache_dir: If not None, `ckpt_dir` should be a shared filesystem path, and this function 
+                            will return as soon as the shards have been saved to this local directory. 
+                            DaemonMoveWorkers will move these shards into `ckpt_dir` in the background.
     """
-    state_dict = to_state_dict(target)
+    # create directories if not exist
     os.makedirs(ckpt_dir, exist_ok=True)
-    ckpt_path = os.path.join(ckpt_dir, f"checkpoint_{step}")
-    with open(ckpt_path, 'wb') as fp:
-        fp.write(
-            msgpack.packb(state_dict,
-                          default=_msgpack_ext_pack_wrapper(ckpt_dir),
-                          strict_types=True))
+    if local_cache_dir is not None:
+        os.makedirs(local_cache_dir, exist_ok=True)
+
+    target = to_state_dict(target)
+    flat_dirs = _dfs_pytree(target, "state")
+    flat_target, target_tree = tree_flatten(target)
+    flat_metadata = []
+    assert(len(flat_dirs) == len(flat_target))
+    for arr_dir, x in zip(flat_dirs, flat_target):
+        arr_path = os.path.join(ckpt_dir, arr_dir)
+        if local_cache_dir is None:
+            arr_cache_path = None
+        else:
+            arr_cache_path = os.path.join(local_cache_dir, arr_dir)
+        if isinstance(x, (DistributedArray, ReplicatedDistributedArray, np.ndarray, jax.xla.DeviceArray)):
+            if isinstance(x, DistributedArray):
+                x.save(arr_path, arr_cache_path)
+            elif isinstance(x, ReplicatedDistributedArray):
+                x.replica.save(arr_path, arr_cache_path)
+            elif isinstance(x, (np.ndarray, jax.xla.DeviceArray)):
+                _save_unsharded_array(arr_path, x)
+            flat_metadata.append(arr_dir)
+        else:
+            flat_metadata.append(x)
+
+    metapath = os.path.join(ckpt_dir, f"checkpoint_{step}")
+    metadata = tree_unflatten(target_tree, flat_metadata)
+    with open(metapath, "wb") as metafile:
+        metafile.write(msgpack.packb(metadata))
```

Comment:
I tried `pickle` first, but it failed to serialize something maybe called `init_func` (I forgot the exact name) in trainState. Since flax uses msgpack in its `save_checkpoint` function, so I use it too.